File: Java/ds/DetectLoopLL.java
import java.util.*;
import java.io.*;
import java.lang.*;
//nodes of Linked list
class Node
{
    int data;
    Node next;
    Node(int x)
    {
        data = x;
        next = null;
    }
}
class DetectLoopLL
{
    public static void makeLoop(Node head, Node tail, int x){
        if (x == 0) return;
        Node curr = head;
        for(int i=1; i<x; i++)
            curr = curr.next;
        tail.next = curr;
    }
    // } Driver Code Start
    public static void main (String[] args){
        Scanner sc = new Scanner(System.in);
        //size of linked list
        System.out.println("\Enter the size of Linked List:");
        int n = sc.nextInt();
         System.out.println("\Enter the data of Linked List:");
        //Input for the Linked list
        int num = sc.nextInt();
        Node head = new Node(num);
        Node tail = head;
        //Linked List created
        for(int i=0; i<n-1; i++)
        {
            num = sc.nextInt();
            tail.next = new Node(num);
            tail = tail.next;
        }
        //Position to check loop exists or not
         System.out.println("\Enter the Position to check:");
        int pos = sc.nextInt();
        makeLoop(head, tail, pos);
        //Object of Solution class
        Solution x = new Solution();
        //True if loop exits otherwise False
        System.out.println("\tOUTPUT");
        System.out.println("-----------------");
        if( x.detectLoop(head) )
            System.out.println("TRUE");
        else
            System.out.println("FALSE");
    }
}
//  Driver Code Ends
//Solution class to check detect the loop
class Solution {
    public static boolean detectLoop(Node head){
        if(head==null||head.next==null)
            return false;
        Node slow = head;
        Node fast = head;
        while(fast != null && fast.next != null) {
            
            slow = slow.next;
            fast = fast.next.next;
            //there exists a loop if and only if 
            //both fast and slow pointer 
            //point to a same one during traversal
            if(slow==fast)
                return true;
        }
        //otherwise return false
        return false;
    }
}
//Sample Input OUTPUT
/*Enter the size of Linked List:
4
Enter the data of Linked List:
1 8 3 4
Enter the Position to check:
0
    OUTPUT
-----------------
FALSE

Time complexity: O(n).
Auxiliary Space:O(1).
*/

File: C/ds/detectLoopLL.c
/*
Detecting a loop/cycle in a linked list.
Algorithm/Approach:
We use two pointers- slow and fast pointers to traverse the linked list.
If there is a loop in the linked list, then we will never encounter a NULL pointer. 
Also, since we are using a slow and a fast pointer, these two pointers will point the same node at one oint of time.
This indicates a loop/cycle in linked list.
*/

#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *next;
} Node;
Node *head;

Node * newNode(int val) {
    Node *new_node = (Node *)malloc(sizeof(Node ));
    new_node -> data = val;
    new_node -> next = NULL;
    return new_node;
}

void insertNode(int value) {
    /* inserting a new node */
    Node *new_node = newNode(value);
    if (!head) {
        head = new_node;
    }
    else {
        new_node -> next = head;
        head = new_node;
    }
}

int detectLoop(Node *head) {
    Node *fastPtr, *slowPtr;
    /* using two pointer */
    slowPtr = fastPtr = head;
    do {
        slowPtr = slowPtr -> next;
        fastPtr = fastPtr -> next;
        fastPtr = fastPtr ? fastPtr -> next : fastPtr;
    } while (fastPtr && slowPtr && fastPtr != slowPtr);
    return slowPtr == fastPtr;
}

void display() {
    Node *p = head;
    while (p) {
        printf("%d ", p -> data);
        p = p -> next;
    }
    printf("\n");
}

int main() {
    int size, val;
    printf("Enter the size of linked list:\n");
    scanf("%d", &size);
    printf("Enter the elements of the linked list:\n");
    for (int i = 0; i < size; i ++) {
        scanf("%d", &val);
        insertNode(val);
    }
    
    /* before introducing a loop */
    if (detectLoop(head)) {
        printf("A loop has been detected in the linked list!\n");
    }
    else {
        printf("No loop detected in the linked list!\n");
    }

    /* after introducing a loop */
    head -> next -> next = head;
    if (detectLoop(head)) {
        printf("A loop has been detected in the linked list!\n");
    }
    else {
        printf("No loop detected in the linked list!\n");
    }
    
    return 0;
}

/*

Sample I/O:
Enter the size of linked list:
5
Enter the elements of the linked list:
1
2
3
4
5
No loop detected in the linked list!
A loop has been detected in the linked list!

Time Complexity: O(n)
Space complexity: O(1)

*/
